src/types.rs

```diff
@@ -540,8 +542,8 @@ fn rewrite_bare_fn(bare_fn: &ast::BareFnTy,
 
     let output = match bare_fn.decl.output {
         FunctionRetTy::Return(ref ty) => Some(&**ty),
-        FunctionRetTy::NoReturn(..) => None,
-        FunctionRetTy::DefaultReturn(..) => unreachable!(),
+        FunctionRetTy::NoReturn(..) => unreachable!(),
```

Comment:
Isn't this reachable?  Example: `const X: fn(i32) -> ! = std::process::exit;`.

Also, this branch has a bunch of commits from #642.